### DISCUSSION TITLE
C libraries with single header AND source file

### DISCUSSION THREAD

**Discussion by acisternino** [edubart/nelua-lang#169] [Q&A] [answered]
Hi,

I am trying to develop, within `nelua-decl`, bindings for [Argtable](https://www.argtable.org/). This library comes as a single *"amalgamated"* C file plus the accompanying header file.  
For reference, similar libraries are `mjson` and `mpc`.

So far I have, amongst others, the following section in the bindings' `.lua` file:

```lua
nldecl.prepend_code = [=[
##[[
cinclude 'argtable3.h'
if not ARGTABLE_NO_IMPL then
  cinclude 'argtable3.c'
end
]]
]=]
```

There are, however, a few things I don't understand:

How am I supposed to handle the C files (both `.c` and `.h`) together in my project?  

I have placed `argtable3.c` and `argtable3.h` in a subdirectory (`clibs`) that I want to be used by the nelua compiler.

I have tried the `-L` option and it didn't work (I think that's only for nelua modules).  
Then I tried `--cflags` and I tried to pass `-Iclibs` but I couldn't find a way to have it accepted by the compiler. It always tries to parse the `-I` itself and bails out with:

```
[foobar]$ nelua --cflags '-I./clibs' main.nelua 
error: unknown option '-I'
Did you mean one of these: '-A' '-B' '-C' '-D' '-H' '-L' '-M' '-P' '-S' '-T' '-V' '-Y' '-a' '-b' '-c' '-d' '-g' '-h' '-i' '-j' '-o' '-q' '-r' '-s' '-t' '-v' '-w'?
```

What is the recommended way of adding local C files to a project?

**Comment by edubart** [comment]
> What is the recommended way of adding local C files to a project?

`## cinclude 'argtable3.c'` should be enough to include a C source file implementation into a Nelua source file, as long they are both in the same directory, but looks like that is already available there, double check if that line exists and is being executed in your `argtable3.nelua` bindings.

Maybe you wanted to add `argtable3` folder to the Nelua modules search path, and using `-L` is correct for that.

In short if you have all 4 files `argtable3.h`, `argtable3.c`, `argtable3.nelua`, `test-argtable3.nelua` in the same folder, and your run `nelua test-argtable3.nelua` from the same folder, you should be able to run without issues. If you create and execute `test-argtable3.nelua` in a different folder, then using `-L` should be enough to expose `argtable3.nelua` and its local dependencies (.h and .c files).

**Comment by acisternino** [comment]
Hi Eduardo

Yes, as long as files are all in the same directory (like in the `nelua-decl` project) things works ok.

My questions were more about what to do in a real project where I want to organize files, modules and libs in different folders.

My original idea was to have this structure:

```
.
├── main.nelua
├── argparse.nelua        <-- unmodified binding file
└── clibs/
    ├── argparse3.c
    └── argparse3.h
```
And this is what didn't work using the `--cflags` option.

I have also tried these two approaches that both worked using `-L`:
```
.
├── main.nelua
└── argparse/             <-- directory named as the module
    ├── init.nelua        <-- this is the binding file renamed
    ├── argparse3.c
    └── argparse3.h
```
and
```
.
├── main.nelua
└── mods/
    ├── argparse.nelua
    ├── argparse3.c
    └── argparse3.h
```

**Comment by edubart** [comment, marked as answer]
In that case, use the following:

```
nelua --cflags='-I./clibs' main.nelua 
```

This should fix the command line error you were getting, notice the `=` added.

**Comment by acisternino** [comment]
It worked like a charm.

Thanks a lot!
